### PR TITLE
Improve scheduler latency.

### PR DESCRIPTION
**Pull request by neunhoef**
This entirely removes direct execution (again!).
At the same time it improves thread handover times to make this bearable.

### CODE REVIEW

**Comment by neunhoef** [comment]
This is code-complete but I would very much like to do some more performance measurements. I will do so in the coming days. Do not merge yet.

**Comment by neunhoef** [comment]
Tests I want to do:
 - [ ] verify single threaded client performance again for single docs and `/_api/version`
 - [ ] verify full throttle load with all threads being busy, check latency 99.9% ile
 - [ ] verify again that index generation cannot block an IO thread (should be trivial)
 - [ ] look at latency under light load with requests spaced out in time a bit

File: arangod/Scheduler/SupervisedScheduler.cpp [hunk 1]
```diff
@@ -266,16 +219,50 @@ bool SupervisedScheduler::queue(RequestLane lane, std::function<void()> handler,
...
-  } else if (sleepyTime_ns > _wakeupTime_ns &&
-             approxQueueLength > _wakeupQueueLength.load(std::memory_order_relaxed)) {
-    doNotify = true;
-  }
-  if (doNotify) {
-    _conditionWork.notify_one();
-  }
+  // PLEASE LEAVE THESE EXPLANATIONS IN THE CODE, SINCE WE HAVE HAD MANY PROBLEMS
+  // IN THIS AREA IN THE PAST, AND WE MIGHT FORGET AGAIN WHAT WE INVESTIGATED IF
+  // IT IS NOT WRITEN ANYWHERE.
```
Comment (jsteemann) [inline arangod/Scheduler/SupervisedScheduler.cpp]:
```suggestion
  // IT IS NOT WRITTEN ANYWHERE.
```

**Comment by jsteemann** [comment]
Nice. Do we have any performance test results of this branch compared to 3.5/previous devel?
Please note that @dothebart is currently testing the HTTP interface performance via arangobench. Is it worth asking him to test this branch too?

**Comment by graetzer** [comment]
why are you using arangobench isn't that using the batch API for most stuff ?

**Comment by graetzer** [comment]
just use https://github.com/arangodb/gobench

**Comment by jsteemann** [comment]
IMHO arangobench doesnt use batching unless explicItly told so

**Comment by jsteemann** [comment]
Just verified this. Default batchSize in arangobench is 0, meaning no batching

**Comment by neunhoef** [comment]
I have now used both gobench and arangobench and confirmed that things are working as I expect them. In particular, the nasty overhead of handing over to a Scheduler thread is basically gone.
Therefore we can now always hand over to a worker thread and do not have to play tricks with direct execution (which invariably breaks other things).

File: arangod/GeneralServer/CommTask.cpp [hunk 2]
```diff
@@ -478,19 +478,21 @@ bool CommTask::handleRequestSync(std::shared_ptr<RestHandler> handler) {
...
   ContentType respType = handler->request()->contentTypeResponse();
   uint64_t mid = handler->messageId();
-  bool handlerAllowDirectExecution = handler->allowDirectExecution();
 
   // queue the operation in the scheduler, and make it eligible for direct execution
   // only if the current CommTask type allows it (HttpCommTask: yes, CommTask: no)
   // and there is currently only a single client handled by the IoContext
-  auto cb = [self = shared_from_this(), handler = std::move(handler)]() mutable {
+  auto cb = [self = shared_from_this(), handler = std::move(handler), a = 0]() mutable {
+    ++a;
```
Comment (maierlars) [inline arangod/GeneralServer/CommTask.cpp]: maybe this should be removed.

File: arangod/Pregel/Algos/SCC.cpp [hunk 3]
```diff
@@ -22,6 +22,7 @@
 
 #include "SCC.h"
 #include <atomic>
+#include <climits>
```
Comment (maierlars) [inline arangod/Pregel/Algos/SCC.cpp]: why?

File: arangod/GeneralServer/CommTask.cpp (hunk 2 above)
Comment (neunhoef) [reply arangod/GeneralServer/CommTask.cpp]: Yes, it should. Will do.

File: arangod/Pregel/Algos/SCC.cpp (hunk 3 above)
Comment (neunhoef) [reply arangod/Pregel/Algos/SCC.cpp]: Was needed for INT_MAX

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/7697/

File: arangod/GeneralServer/CommTask.cpp (hunk 2 above)
Comment (neunhoef) [reply arangod/GeneralServer/CommTask.cpp]: Done.

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/7705/

**Comment by neunhoef** [comment]
This should now also fix the Windows compilation issues.

**Comment by neunhoef** [comment]
Vadim confirmed that the one test failure observed in the latest run is an unstable test which happens occasionally. I will merge this now.

**Pull request merged**